### PR TITLE
adding `DeploymentName` for alerts metadata

### LINKED ISSUE #1065: Addition of Deployment Name in KubeArmor Alerts/Telemetry And Summarized Observability
## Feature Request

**Short Description**

In addition to other metadata like pod name, namespace, and cluster name, having the **Deployment Name** for Alerts and telemetry, and Summarised Observability will give more visibility. 

Tasks:
- [x] Handle deployments
- [x] Handle replicasets
- [x] Handle statefulsets
- [x] Handle daemonset
- [x] Check what other things needs to be handled
- [x] Add testcase for each

I suggest telemetry format which is like (note the nested json):
```
...
"owner": {
    "ref": "deployments/daemonsets/replicasets/statefulsets",
    "name": "ubunutu-1-deploy",
    "namespace: "default",
}
...
```

### PR DESCRIPTION
**Pull request by Prateeknandle**
**Purpose of PR?**:

Fixes #1065 

**Description:**
Adding owner information of the pod for alerts metadata

**Does this PR introduce a breaking change?**
No

**If the changes in this PR are manually verified, list down the scenarios covered:**:
tested for each controller - deployments, replicaset, statefulset, daemonset

**Output:**
```
{
  "Timestamp": 1679588263,
  "UpdatedTime": "2023-03-23T16:17:43.146109Z",
  "ClusterName": "default",
  "HostName": "prateek-lenovo-ideapad-310-15isk",
  "NamespaceName": "default",
  "Owner": {
    "Ref": "DaemonSet",
    "Name": "my-daemonset",
    "Namespace": "default"
  },
  "PodName": "my-daemonset-xhrr9",
  "Labels": "app=my-app",
  "ContainerID": "0e6a98ec00521ed92fd29664ae238f60463a93b819e6c54a4494cf371e63e86a",
  "ContainerName": "my-container",
  "ContainerImage": "docker.io/library/nginx:latest@sha256:1ed4dff2b99011798f2c228667d7cb4f4e2bd76b2adc78fd881d39f923e78c9d",
  "HostPPID": 671928,
  "HostPID": 672009,
  "PPID": 41,
  "PID": 46,
  "ParentProcessName": "/bin/bash",
  "ProcessName": "/bin/sleep",
  "PolicyName": "new",
  "Severity": "5",
  "Message": "block /bin/sleep",
  "Type": "MatchedPolicy",
  "Source": "/bin/bash",
  "Operation": "Process",
  "Resource": "/bin/sleep",
  "Data": "syscall=SYS_EXECVE",
  "Enforcer": "AppArmor",
  "Action": "Block",
  "Result": "Permission denied"
}
```

### CODE REVIEW

**Comment by vishnusomank** [comment]
What will happen if the pod is not deployed using a Deployment file? Will it show empty data or the field will be removed? @Prateeknandle

**Comment by Prateeknandle** [comment]
> What will happen if the pod is not deployed using a Deployment file? Will it show empty data or the field will be removed? @Prateeknandle

field will not be present @vishnusomank

**Review by vishnusomank — APPROVED** [review]
LGTM!

**Comment by Prateeknandle** [comment]
@achrefbensaad @vishnusomank can we merge this
@achrefbensaad  do you have any review comments to add?

File: KubeArmor/core/kubeUpdate.go [hunk 1]
```diff
@@ -190,6 +190,7 @@ func (dm *KubeArmorDaemon) UpdateEndPointWithPod(action string, pod tp.K8sPod) {
 
 		newPoint.NamespaceName = pod.Metadata["namespaceName"]
 		newPoint.EndPointName = pod.Metadata["podName"]
+		newPoint.DeploymentName = pod.Metadata["deploymentName"]
```
Comment (nyrahul) [inline KubeArmor/core/kubeUpdate.go]: We need to handle deployments, replicasets, statefulsets etc.
Currently, only deployment name is handled. I think the issue explains possible way to handle it.

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/core/kubeUpdate.go (hunk 1 above)
Comment (Prateeknandle) [reply KubeArmor/core/kubeUpdate.go]: updated the pr with owner information

**Comment by Prateeknandle** [comment]
@rksharma95 can you please review the pr

**Comment by Prateeknandle** [comment]
Should we remove `NmaespaceName` from the metadata ?

**Comment by rksharma95** [comment]
> Should we remove `NmaespaceName` from the metadata ?

i think we can keep it, it looks like a duplication but it's obviously differentiable to map each namespace. i.e. one for pod itself and one for owner controller. WDYT?

**Review by rksharma95 — APPROVED** [review]
LGTM :+1:

File: KubeArmor/core/kubeUpdate.go [hunk 2]
```diff
@@ -190,6 +190,9 @@ func (dm *KubeArmorDaemon) UpdateEndPointWithPod(action string, pod tp.K8sPod) {
 
 		newPoint.NamespaceName = pod.Metadata["namespaceName"]
 		newPoint.EndPointName = pod.Metadata["podName"]
+		newPoint.Owner.Ref = pod.Metadata["owner.controller"]
+		newPoint.Owner.Name = pod.Metadata["owner.controllerName"]
+		newPoint.Owner.Namespace = pod.Metadata["namespaceName"]
```
Comment (nyrahul) [inline KubeArmor/core/kubeUpdate.go]: This namespace needs to be picked from `owner`.

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: KubeArmor/core/kubeUpdate.go (hunk 2 above)
Comment (Prateeknandle) [reply KubeArmor/core/kubeUpdate.go]: updated

**Comment by nyrahul** [comment]
Thank you @stefin9898 for your interest.
Can you please test this PR and check if it works?

**Comment by stefin9898** [comment]
Hi,
I checked the kubearmor logs in `tmp/kubearmor.log` 

<details>
<summary>For Deployment action: Block</summary>

```json
{
  "timestamp": 1682064422,
  "updatedTime": "2023-04-21T08:07:02.425399Z",
  "hostName": "dell-latitude",
  "namespaceName": "multiubuntu",
  "owner": {
    "ref": "Deployment",
    "name": "ubuntu-4-deployment",
    "namespace": "multiubuntu"
  },
  "podName": "ubuntu-4-deployment-765d67b865-56tk8",
  "labels": "container=ubuntu-4,group=group-2",
  "containerID": "210e7fb2e4bfcc3de2092b369f63e477b339df5e1998951fb29531ad11e44285",
  "containerName": "ubuntu-4-container",
  "containerImage": "docker.io/kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260",
  "hostPPid": 60140,
  "hostPid": 60143,
  "ppid": 60140,
  "pid": 200,
  "uid": 0,
  "parentProcessName": "",
  "processName": "/bin/bash",
  "enforcer": "eBPF Monitor",
  "policyName": "DefaultPosture",
  "atags": [],
  "type": "MatchedPolicy",
  "source": "/bin/bash",
  "operation": "File",
  "resource": "/lib/x86_64-linux-gnu/libtinfo.so.5.9",
  "data": "syscall=SYS_OPENAT fd=-100 flags=O_RDONLY|O_CLOEXEC",
  "action": "Block",
  "result": "Permission denied"
}
```
</details>
<details>
<summary>For Replicaset action: Block</summary>

```json
  {
  "timestamp": 1682065652,
  "updatedTime": "2023-04-21T08:27:32.181905Z",
  "hostName": "dell-latitude",
  "namespaceName": "multiubuntu",
  "owner": {
    "ref": "ReplicaSet",
    "name": "ubuntu-1-deployment",
    "namespace": "multiubuntu"
  },
  "podName": "ubuntu-1-deployment-v9lrg",
  "labels": "container=ubuntu-1,group=group-1",
  "containerID": "53b1239bb9a27a1179def3841d06e7f973afbd896b81043582298c1ebd74f223",
  "containerName": "ubuntu-1-container",
  "containerImage": "docker.io/kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260",
  "hostPPid": 81125,
  "hostPid": 81130,
  "ppid": 81125,
  "pid": 112,
  "uid": 0,
  "parentProcessName": "",
  "processName": "/bin/bash",
  "enforcer": "eBPF Monitor",
  "policyName": "DefaultPosture",
  "atags": [],
  "type": "MatchedPolicy",
  "source": "/bin/bash",
  "operation": "File",
  "resource": "/lib/x86_64-linux-gnu/libtinfo.so.5.9",
  "data": "syscall=SYS_OPENAT fd=-100 flags=O_RDONLY|O_CLOEXEC",
  "action": "Block",
  "result": "Permission denied"
}
```
</details>
<details>
<summary>For Statefulset action: Block</summary>
  
```json
 {
  "timestamp": 1682065853,
  "updatedTime": "2023-04-21T08:30:53.458003Z",
  "hostName": "dell-latitude",
  "namespaceName": "multiubuntu",
  "owner": {
    "ref": "StatefulSet",
    "name": "ubuntu-1-deployment",
    "namespace": "multiubuntu"
  },
  "podName": "ubuntu-1-deployment-0",
  "labels": "container=ubuntu-1,group=group-1,statefulset.kubernetes.io/pod-name=ubuntu-1-deployment-0",
  "containerID": "bea7438ee546b6fb7bf21d9855b7e024481e2cad92c1ee7d81dd9e694a018c0c",
  "containerName": "ubuntu-1-container",
  "containerImage": "docker.io/kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260",
  "hostPPid": 84950,
  "hostPid": 84953,
  "ppid": 84950,
  "pid": 101,
  "uid": 0,
  "parentProcessName": "",
  "processName": "/bin/bash",
  "enforcer": "eBPF Monitor",
  "policyName": "DefaultPosture",
  "atags": [],
  "type": "MatchedPolicy",
  "source": "/bin/bash",
  "operation": "File",
  "resource": "/lib/x86_64-linux-gnu/libtinfo.so.5.9",
  "data": "syscall=SYS_OPENAT fd=-100 flags=O_RDONLY|O_CLOEXEC",
  "action": "Block",
  "result": "Permission denied"
}
```
</details>
<details>
<summary>For Daemonset action: Block</summary>
  
```json
 {
  "timestamp": 1682066071,
  "updatedTime": "2023-04-21T08:34:31.986474Z",
  "hostName": "dell-latitude",
  "namespaceName": "multiubuntu",
  "owner": {
    "ref": "DaemonSet",
    "name": "ubuntu-1-deployment",
    "namespace": "multiubuntu"
  },
  "podName": "ubuntu-1-deployment-nw657",
  "labels": "container=ubuntu-1,group=group-1",
  "containerID": "d38540005caf92018484509269a330049159531e438a14bab6f4328ed4a111cc",
  "containerName": "ubuntu-1-container",
  "containerImage": "docker.io/kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260",
  "hostPPid": 88935,
  "hostPid": 88941,
  "ppid": 88935,
  "pid": 101,
  "uid": 0,
  "parentProcessName": "",
  "processName": "/bin/bash",
  "enforcer": "eBPF Monitor",
  "policyName": "DefaultPosture",
  "atags": [],
  "type": "MatchedPolicy",
  "source": "/bin/bash",
  "operation": "File",
  "resource": "/lib/x86_64-linux-gnu/libtinfo.so.5.9",
  "data": "syscall=SYS_OPENAT fd=-100 flags=O_RDONLY|O_CLOEXEC",
  "action": "Block",
  "result": "Permission denied"
}
```
</details>
<details>
<summary>For General events</summary>
  
```json
{
  "timestamp": 1682065588,
  "updatedTime": "2023-04-21T08:26:28.700251Z",
  "hostName": "dell-latitude",
  "owner": {
    "ref": "",
    "name": "",
    "namespace": ""
  },
  "hostPPid": 55841,
  "hostPid": 80064,
  "ppid": 55841,
  "pid": 80064,
  "uid": 1000,
  "parentProcessName": "/usr/bin/zsh",
  "processName": "/usr/bin/grep",
  "atags": null,
  "type": "HostLog",
  "source": "/usr/bin/grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn --exclude-dir=.idea --exclude-dir=.tox denied",
  "operation": "File",
  "resource": "/usr/share/locale/locale.alias",
  "data": "syscall=SYS_OPENAT fd=-100 flags=O_RDONLY|O_CLOEXEC",
  "result": "Passed"
}
```
</details>

Note: These are from kubarmor log file not from `karmor logs`.

@nyrahul

**Comment by nyrahul** [comment]
Can you add an automated test that can test the above events?

**Comment by nyrahul** [comment]
For general events, currently we are sending empty fields ... it is better we omit the owner sub structure from the json itself. @stefin9898 

```json
 "owner": {
    "ref": "",
    "name": "",
    "namespace": ""
  },
```

**Comment by stefin9898** [comment]
@nyrahul Is this the struct you had in mind?
```json
{
  "timestamp": 1682322414,
  "updatedTime": "2023-04-24T07:46:54.404253Z",
  "hostName": "dell-latitude",
  "owner": {},
  "hostPPid": 279232,
  "hostPid": 339569,
  "ppid": 279232,
  "pid": 339569,
  "uid": 1000,
  "parentProcessName": "/usr/bin/zsh",
  "processName": "/usr/bin/kubectl",
  "atags": null,
  "type": "HostLog",
  "source": "/usr/bin/kubectl",
  "operation": "Network",
  "resource": "remoteip=127.0.0.1 port=8001 protocol=TCP",
  "data": "kprobe=tcp_accept domain=AF_INET",
  "result": "Unknown error"
}
```

Edit: the test suites are ready waiting for this PR to be merged

**Comment by nyrahul** [comment]
> ```json
> {
>   "timestamp": 1682322414,
>   "updatedTime": "2023-04-24T07:46:54.404253Z",
>   "hostName": "dell-latitude",
>   "owner": {},
>   "hostPPid": 279232,
>   "hostPid": 339569,
>   "ppid": 279232,
>   "pid": 339569,
>   "uid": 1000,
>   "parentProcessName": "/usr/bin/zsh",
>   "processName": "/usr/bin/kubectl",
>   "atags": null,
>   "type": "HostLog",
>   "source": "/usr/bin/kubectl",
>   "operation": "Network",
>   "resource": "remoteip=127.0.0.1 port=8001 protocol=TCP",
>   "data": "kprobe=tcp_accept domain=AF_INET",
>   "result": "Unknown error"
> }
> ```

I am expecting that the owner field itself is not present.
```json
{
  "timestamp": 1682322414,
  "updatedTime": "2023-04-24T07:46:54.404253Z",
  "hostName": "dell-latitude",
  "hostPPid": 279232,
  "hostPid": 339569,
  "ppid": 279232,
  "pid": 339569,
  "uid": 1000,
  "parentProcessName": "/usr/bin/zsh",
  "processName": "/usr/bin/kubectl",
  "atags": null,
  "type": "HostLog",
  "source": "/usr/bin/kubectl",
  "operation": "Network",
  "resource": "remoteip=127.0.0.1 port=8001 protocol=TCP",
  "data": "kprobe=tcp_accept domain=AF_INET",
  "result": "Unknown error"
}
```

**Comment by stefin9898** [comment]
@nyrahul Changes are added to the latest commit in this branch 

Now for general events

**Comment by stefin9898** [comment]
Changes are added to the latest commit in this PR @nyrahul 
<details>
<summary>For Statefulset action: Block</summary>

```json
{
  "timestamp": 1682414463,
  "updatedTime": "2023-04-25T09:21:03.377704Z",
  "hostName": "dell-latitude",
  "namespaceName": "multiubuntu",
  "owner": {
    "ref": "StatefulSet",
    "name": "ubuntu-1-deployment-statefulset",
    "namespace": "multiubuntu"
  },
  "podName": "ubuntu-1-deployment-statefulset-0",
  "labels": "group=group-1,statefulset.kubernetes.io/pod-name=ubuntu-1-deployment-statefulset-0,container=ubuntu-1",
  "containerID": "5e64a6a1547f65c9f640b634ac639b67f0e53200bb50be09e09f12c4b9171644",
  "containerName": "ubuntu-1-container",
  "containerImage": "docker.io/kubearmor/ubuntu-w-utils:0.1@sha256:b4693b003ed1fbf7f5ef2c8b9b3f96fd853c30e1b39549cf98bd772fbd99e260",
  "hostPPid": 973186,
  "hostPid": 973413,
  "ppid": 114,
  "pid": 125,
  "uid": 0,
  "parentProcessName": "/bin/bash",
  "processName": "/bin/sleep",
  "enforcer": "BPFLSM",
  "policyName": "ksp-group-1-proc-path-block",
  "severity": "1",
  "atags": null,
  "type": "MatchedPolicy",
  "source": "/bin/bash",
  "operation": "Process",
  "resource": "/bin/sleep",
  "data": "syscall=SYS_EXECVE",
  "action": "Block",
  "result": "Permission denied"
}
```
</details>
<details>
<summary>For general events </summary>

```json
{
  "timestamp": 1682414457,
  "updatedTime": "2023-04-25T09:20:57.645885Z",
  "hostName": "dell-latitude",
  "hostPPid": 973277,
  "hostPid": 973345,
  "ppid": 973277,
  "pid": 973345,
  "uid": 1000,
  "parentProcessName": "/usr/bin/zsh",
  "processName": "/usr/bin/grep",
  "atags": null,
  "type": "HostLog",
  "source": "/usr/bin/grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn --exclude-dir=.idea --exclude-dir=.tox \"/bin/sleep",
  "operation": "File",
  "resource": "/usr/lib/libc.so.6",
  "data": "syscall=SYS_OPENAT fd=-100 flags=O_RDONLY|O_CLOEXEC",
  "result": "Passed"
}
```
</details>

Note: These are from kubarmor log file not from `karmor logs`.

**Review by nyrahul — APPROVED** [review]
LGTM

**Pull request merged**